nidx/nidx_paragraph/src/search_query.rs

```diff
@@ -321,62 +152,63 @@ pub fn suggest_query(
-    search: &ParagraphSearchRequest,
+    request: &ParagraphSearchRequest,
     prefilter: &PrefilterResult,
+    index: &Index,
     schema: &ParagraphSchema,
-    distance: u8,
-    with_advance: Option<Box<dyn Query>>,
 ) -> (Box<dyn Query>, SharedTermC, Box<dyn Query>) {
-    let mut term_collector = TermCollector::default();
-    let processed = preprocess_raw_query(text, &mut term_collector);
-    let query = parse_query(parser, &processed.regular_query);
-    let fuzzy_query = parse_query(parser, &processed.fuzzy_query);
-    let termc = SharedTermC::from(term_collector);
-    let mut fuzzies = fuzzied_queries(fuzzy_query, false, distance, termc.clone());
-    let mut originals = vec![(Occur::Must, query)];
-    if let Some(advance) = with_advance {
-        originals.push((Occur::Must, advance.box_clone()));
-        fuzzies.push((Occur::Must, advance));
+    let query = &request.body;
+    let parser_config = ParserConfig::with_schema(schema);
+    let parsed = query_parser::parse_query(query, parser_config);
+    let ParsedQuery {
+        keyword,
+        fuzzy,
+        term_collector,
+    } = parsed;
+
+    let mut keyword_subqueries = vec![(Occur::Must, keyword)];
+    let mut fuzzy_subqueries = vec![(Occur::Must, fuzzy)];
+
+    if let Some(advanced_query) = &request.advanced_query {
+        // The advanced query allow users to leverage tantivy grammar in our system. We are not
+        // interested in failing here, so we use the lenient parser and ignore errors
+        let parser = QueryParser::for_index(index, vec![schema.text]);
+        let (advanced_query, _errors) = parser.parse_query_lenient(advanced_query);
+
+        keyword_subqueries.push((Occur::Must, advanced_query.box_clone()));
+        fuzzy_subqueries.push((Occur::Must, advanced_query));
     }
 
-    let filter_query = filter_query(schema, prefilter, &search.filtering_formula, search.filter_or);
+    let filter_query = filter_query(schema, prefilter, &request.filtering_formula, request.filter_or);
     if let Some(query) = filter_query {
-        originals.push((Occur::Must, query.box_clone()));
-        fuzzies.push((Occur::Must, query));
+        keyword_subqueries.push((Occur::Must, query.box_clone()));
+        fuzzy_subqueries.push((Occur::Must, query));
     }
 
-    if !search.with_duplicates {
+    if !request.with_duplicates {
         let term = Term::from_field_u64(schema.repeated_in_field, 0);
         let term_query = TermQuery::new(term, IndexRecordOption::Basic);
-        fuzzies.push((Occur::Must, Box::new(term_query.clone())));
-        originals.push((Occur::Must, Box::new(term_query)))
+        fuzzy_subqueries.push((Occur::Must, Box::new(term_query.clone())));
+        keyword_subqueries.push((Occur::Must, Box::new(term_query)))
     }
 
-    if originals.len() == 1 && originals[0].1.is::<AllQuery>() {
-        let original = originals.pop().unwrap().1;
+    if keyword_subqueries.len() == 1 && keyword_subqueries[0].1.is::<AllQuery>() {
```

Comment:
Do you know why is this here and if it's needed?

Reply:
Not sure...

Reply:
I think it was a weird combination of things that could happen. An `AllQuery` was the fallback for queries failing tantivy's parsing. So, IMO it was a shortcut like: if we can't parse a query and use an `AllQuery`, fuzzy won't be needed so use an empty query. But in a weird fashion